### ISSUE TITLE
Decoupling faircoin (new app)

### ISSUE THREAD

**Issue by XaviP** [FreedomCoop/valuenetwork#290]
Trying to find a doable road map. A (very schematic and simple) proposal:

1. Decoupling methods 
- Move all faircoin functions/methods from valueaccounting/models.py and views.py inside services.py (class ExchangeService)
- Change all references to those functions/methods in valueaccouting and work apps to services.py
- Testing and deploy.

2. Decoupling fields
- Create a new faircoin app with a model with the fields used in valueaccounting/models.py with OneToOne relation with the objects which contain those fields.
- Copy from production db the content in those fields in valueaccounting app to the faircoin app model new fields.
- Change all references to those fields to the new ones in faircoin app.
- Testing and deploy.

3. Make the app pluggable 
- Move valueaccounting/services.py to faircoin app
- Use conditionals for importing and use anything from faircoin app ( if 'faircoin' in settings.INSTALLED_APPS)

Please, contribute to improve it (it can be edited).
Where are the black holes in this road map?

**Comment by XaviP** [comment]
- @bhaugen said "just decoupling the methods, and leaving the fields where they were, so that (for example,) an agent could have several resources representing accounts in different currencies with for example different banks and blockchains, and not factoring out separate objects for example for the digital currency fields on EconomicEvent"
- I think @escanda was proposing to use the methods in a generic way too (with URIs or an interface) for all currencies.

**Comment by XaviP** [comment]
- @escanda said "I'd create an app called wallet with an interface.py file with an ABC class detailing the methods and properties to be declared in subclasses, a __init__.py file containing the factory function, and then multiple files, each for each coin type implementing interface.py 's class. In models.py I'd have a many2one rel as a STA (single table inheritance) through https://github.com/craigds/django-typed-models then each delegate or backend, could choose its object from the list and perform the related operations"

**Comment by XaviP** [comment]
Initial commit wallet app: https://github.com/FreedomCoop/valuenetwork/commit/29449692a9561bb4c421c6c0f45b73f3e352497b
Two notes:
- I would like to begin this app with compatibility python 2&3, but I don't find the way to make it using abc package for the abstract class.
- Please, review the vocabulary used, I think it's important to be precise. "gateway" is the word I found to define what we are doing, but feel free to change anything.

**Comment by XaviP** [comment]
Related to this issue, working in faircoin decoupling in branch new_faircoin_app, with this main gaols:
- Moving from work app views to faircoin app: faircoin_account, faircoin_history, and all stuff related (ajax views)
- Moving from work app templates to faircoin app: templates related to views explained before.
- Moving electrum_fair_nrp.py to faircoin.utils
- Moving daemon sample files inside faircoin app.
- Moving all commands (cron) related to faircoin to faircoin app.
This way, all related to faircoin will be in a separated app.

**Comment by XaviP** [comment]
The branch new_faircoin_app is ready to test in testocp. The wallet is working ok, but to test further than faircoin_account and faircoin_history pages is needed, as I don't really know the implications inside valueaccounting and work apps that these movements of code have.
Another change done: valueaccounting/faircoin_utils.py is merged into faircoin/utils.py, and all imports in valueaccountig and work apps are changed to point there.

**Comment by XaviP** [comment]
Transfering specific to faircoin stuff from valueaccounting/models.py to faircoin/models.py:
https://github.com/FreedomCoop/valuenetwork/blob/faircoin_models/faircoin/models.py
It's not working yet, it's only a draft.

@bhaugen, can you take a look? Am I missing something? Do you catch my purpose? Do you agree?

**Comment by bhaugen** [comment]
I took a look. I think I catch the purpose: FaircoinAddress is oneToOne with EconomicResource, which is where faircoin addresses have been held before. So it is factoring out all the faircoin-specific behaviors from the valueaccounting app, and then you will be able to add more behaviors for different currencies behind other gateways.

FaircoinTransaction does the same for EconomicEvent.

Is that what you had in mind? 

I didn't see any corresponding changes in https://github.com/FreedomCoop/valuenetwork/blob/faircoin_models/valuenetwork/valueaccounting/models.py though. Are those yet to come?

The first part makes sense to me, if the corresponding changes in valueaccounting and work apps are smooth enough.

**Comment by XaviP** [comment]
> I didn't see any corresponding changes in https://github.com/FreedomCoop/valuenetwork/blob/faircoin_models/valuenetwork/valueaccounting/models.py though. Are those yet to come?

Yes, when everything goes fine, I'll erase the fields/functions from valueaccounting

> The first part makes sense to me, if the corresponding changes in valueaccounting and work apps are smooth enough.
Yes, I'll try that smoothness :-)

**Comment by bhaugen** [comment]
>  I'll erase the fields/functions from valueaccounting

Won't you need to add some new functions (or redo some of those that exist), at least for EconomicAgent? (But I'll wait and see what comes next.)

**Comment by XaviP** [comment]
Hey, too complex, so I'm going to work in two fases:
1. Add new faircoin models and make it work (with transvase of data in db to new models)
- This step is passing tests in branch:
https://github.com/FreedomCoop/valuenetwork/tree/faircoin_models_onlyfields
- Needs more test and deployment.
2. Move methods and references to them.

**Comment by bhaugen** [comment]
> Find a secure way to migrate db data from valueaccounting app tables to faircoin app new tables in testocp and production dbs.

[Data migrations](https://docs.djangoproject.com/en/1.11/topics/migrations/#data-migrations) might work.

**Comment by XaviP** [comment]
- New fields created in faircoin models.py
- Migration data file included (from EconomicResource and EconomicEvent to faircoin models.py)
- Migrated data in production and code working ok with new fields.
- Still to delete old fields in production.

**Comment by XaviP** [comment]
- Old fields deleted in production. Everything works ok.
- Next step should be move all models functions related to faircoin from EconomicAgent, EconomicResource and EconomicEvent to FaircoinAddress and FaircoinTransaction models. And move valueaccounting/services.py to faircoin app.
- This should finish with faircoin decoupling. Later, to create a new gateway app is needed in order to abstract the interface between rea core and all gateways implemented. There's a draft in wallet_app branch:
https://github.com/FreedomCoop/valuenetwork/tree/wallet_app/wallet
